File: backend/Athkar/Areas/Domain/Widgets/WidgetSettings.cs
using Athkar.Areas.Domain.Content;
using Athkar.Shareds.Enums;
using Athkar.Shareds.Models.Base;

namespace Athkar.Areas.Domain.Widgets;

/// <summary>
/// What the admin decides about home-screen widgets.
///
/// Exactly one live row, like <see cref="Configuration.AppConfiguration"/> —
/// the seeder creates it and the service only updates it.
///
/// The reason this is admin-owned at all: a widget is the one surface of this
/// app that a reader sees **without opening it**, so it is also the one whose
/// content nobody can correct in the moment. An admin who has withdrawn a
/// disputed dhikr needs it gone from the home screen too, and the only way to
/// reach that is a setting the app syncs.
///
/// What it deliberately does *not* carry: the text itself. The widget draws
/// from the catalogue already on the device, so it keeps working with the radio
/// off — these are the rules for choosing, not the content.
/// </summary>
public class WidgetSettings : AuditableEntity
{
    /// <summary>
    /// The master switch. Off hides the widget from the app's editor and makes
    /// the native widget render a quiet placeholder rather than stale content.
    /// </summary>
    public bool IsEnabled { get; set; } = true;

    /// <summary>What a reader gets before they choose anything.</summary>
    public WidgetKind DefaultKind { get; set; } = WidgetKind.NextPrayer;

    /// <summary>
    /// Which kinds the reader may pick. Both false is a valid, if severe,
    /// configuration — it is the same as switching widgets off, and the service
    /// says so rather than silently correcting it.
    /// </summary>
    public bool AllowPrayerWidget { get; set; } = true;

    public bool AllowDhikrWidget { get; set; } = true;

    public WidgetTheme Theme { get; set; } = WidgetTheme.System;

    /// <summary>
    /// How often the app refreshes the widget's values, in minutes.
    ///
    /// Clamped rather than free: under fifteen minutes Android coalesces the
    /// updates anyway and the only measurable effect is battery, which is the
    /// complaint a widget earns fastest.
    /// </summary>
    public int RefreshMinutes { get; set; } = 30;

    public bool ShowHijriDate { get; set; } = true;

    /// <summary>Whether the prayer widget counts down or just names the time.</summary>
    public bool ShowCountdown { get; set; } = true;

    /// <summary>
    /// Which chapter the dhikr widget draws from.
    ///
    /// Null means "follow the time of day", which is the good default and what
    /// the home screen already does. Setting it pins the widget to one chapter
    /// — how a Ramadan plan or a themed month is run without an app update.
    /// </summary>
    public int? DhikrCategoryId { get; set; }
    public AthkarCategory? DhikrCategory { get; set; }

    /// <summary>
    /// The gallery entry a reader gets before they have chosen one — matched
    /// against <see cref="WidgetCatalogItem.Key"/>.
    ///
    /// A key rather than a foreign key, deliberately. The app resolves it
    /// against its own renderer registry, and a build that has never heard of
    /// this key must fall back rather than fail; a relational reference would
    /// imply a guarantee the app cannot honour. Null means "the first entry the
    /// reader's app can actually draw", which is the honest default for an
    /// install that may be a release behind this server.
    /// </summary>
    public string? DefaultWidgetKey { get; set; }

    // ── What the reader is allowed to change for themselves ──
    //
    // These are permissions, not preferences: the reader's own choice lives on
    // their phone and never reaches this server. They are here because a
    // customisation can break the one promise the widget makes — a transparent
    // background over a photographic wallpaper can render a dhikr unreadable,
    // and an unreadable dhikr on a home screen is worse than no widget — so the
    // project keeps a way to withdraw one without shipping an app update.

    /// <summary>Whether the reader may tint the widget's background.</summary>
    public bool AllowBackgroundColor { get; set; } = true;

    /// <summary>Whether the reader may drop the panel and let the wallpaper through.</summary>
    public bool AllowTransparency { get; set; } = true;

    /// <summary>Whether the reader may put one of their own photographs behind it.</summary>
    public bool AllowBackgroundImage { get; set; } = true;

    /// <summary>
    /// Whether the reader may pin their own text — an ayah, a du'a, a line they
    /// want in front of them all day.
    ///
    /// The one place in this app where a widget shows words that carry no
    /// takhrij. It stays honest because the text is the reader's own, typed on
    /// their own phone, never uploaded and never shown to anyone else; the app
    /// labels it as theirs rather than dressing it as catalogue content.
    /// </summary>
    public bool AllowCustomWidget { get; set; } = true;

    /// <summary>
    /// How much of their own text the reader may pin. A widget is a few square
    /// centimetres, so this is a legibility limit before it is a storage one.
    /// </summary>
    public int CustomWidgetMaxLength { get; set; } = 280;

    /// <summary>
    /// Bumped on every change, so a device can tell whether what it last pushed
    /// to the launcher is still current without diffing every field.
    /// </summary>
    public int Version { get; set; } = 1;
}

File: app/athkar_app/lib/models/widget_settings.dart
/// The admin's rules for the home-screen widget.
///
/// Mirrors `Areas/Domain/Widgets/WidgetSettings.cs`. Cached on the device like
/// everything else the app needs at launch — the widget has to keep drawing
/// with the radio off.
library;

/// Mirrors `Shareds/Enums/WidgetKind.cs`.
enum WidgetKind {
  nextPrayer(1),
  dhikr(2),
  combined(3);

  const WidgetKind(this.value);
  final int value;

  static WidgetKind fromValue(int? value) => switch (value) {
        2 => WidgetKind.dhikr,
        3 => WidgetKind.combined,
        _ => WidgetKind.nextPrayer,
      };

  static WidgetKind fromName(String? name) {
    for (final kind in WidgetKind.values) {
      if (kind.name == name) return kind;
    }
    return WidgetKind.nextPrayer;
  }
}

/// Mirrors `Shareds/Enums/WidgetTheme.cs`.
enum WidgetTheme {
  system(0),
  light(1),
  dark(2),
  transparent(3);

  const WidgetTheme(this.value);
  final int value;

  static WidgetTheme fromValue(int? value) => switch (value) {
        1 => WidgetTheme.light,
        2 => WidgetTheme.dark,
        3 => WidgetTheme.transparent,
        _ => WidgetTheme.system,
      };
}

class WidgetSettings {
  const WidgetSettings({
    required this.isEnabled,
    required this.defaultKind,
    required this.allowPrayerWidget,
    required this.allowDhikrWidget,
    required this.theme,
    required this.refreshMinutes,
    required this.showHijriDate,
    required this.showCountdown,
    required this.allowBackgroundColor,
    required this.allowTransparency,
    required this.allowBackgroundImage,
    required this.allowCustomWidget,
    required this.customWidgetMaxLength,
    required this.version,
    this.defaultWidgetKey,
    this.dhikrCategoryId,
  });

  final bool isEnabled;
  final WidgetKind defaultKind;
  final bool allowPrayerWidget;
  final bool allowDhikrWidget;
  final WidgetTheme theme;
  final int refreshMinutes;
  final bool showHijriDate;
  final bool showCountdown;

  // What the reader is allowed to change for themselves.
  //
  // Permissions, not preferences: the reader's own choice lives here on the
  // phone and never reaches the server. They exist because a customisation can
  // break the one promise a widget makes — a transparent panel over a
  // photographic wallpaper can render a dhikr unreadable, and an unreadable
  // dhikr on a home screen is worse than no widget — so the project keeps a way
  // to withdraw one without shipping an app update.
  final bool allowBackgroundColor;
  final bool allowTransparency;
  final bool allowBackgroundImage;

  /// Whether the reader may pin their own text — an ayah, a du'a, a line.
  ///
  /// The one place in this app where a widget shows words that carry no
  /// takhrij. It stays honest because the text is the reader's own, typed on
  /// their own phone, never uploaded and never shown to anyone else, and the
  /// app labels it as theirs rather than dressing it as catalogue content.
  final bool allowCustomWidget;

  final int customWidgetMaxLength;

  /// The gallery entry a reader gets before they have chosen one.
  ///
  /// A key, matched against this build's renderer registry. Null — and an
  /// unknown key — both mean the same thing here: fall back to the first entry
  /// this build can actually draw. The server may be a release ahead of the app,
  /// so an unrecognised default has to degrade rather than leave a home screen
  /// blank.
  final String? defaultWidgetKey;

  /// Null means the widget follows the time of day, as the home screen does.
  final int? dhikrCategoryId;

  final int version;

  /// What a first launch uses before the server has answered. Matches the
  /// server's own defaults, so nothing changes shape when it does.
  static const fallback = WidgetSettings(
    isEnabled: true,
    defaultKind: WidgetKind.nextPrayer,
    allowPrayerWidget: true,
    allowDhikrWidget: true,
    theme: WidgetTheme.system,
    refreshMinutes: 30,
    showHijriDate: true,
    showCountdown: true,
    allowBackgroundColor: true,
    allowTransparency: true,
    allowBackgroundImage: true,
    allowCustomWidget: true,
    customWidgetMaxLength: 280,
    version: 0,
  );

  /// The kinds the reader may actually pick, in a stable order.
  List<WidgetKind> get allowedKinds => [
        if (allowPrayerWidget) WidgetKind.nextPrayer,
        if (allowDhikrWidget) WidgetKind.dhikr,
      ];

  /// The reader's choice, corrected to one the admin still allows.
  ///
  /// This is what makes withdrawing a kind actually reach a home screen: a
  /// reader who picked the dhikr widget last month falls back to the prayer one
  /// rather than keeping something the admin has taken away.
  WidgetKind resolveKind(WidgetKind preferred) {
    final allowed = allowedKinds;
    if (allowed.isEmpty) return defaultKind;
    if (allowed.contains(preferred)) return preferred;

    return allowed.contains(defaultKind) ? defaultKind : allowed.first;
  }

  factory WidgetSettings.fromJson(Map<String, dynamic> json) => WidgetSettings(
        isEnabled: json['isEnabled'] as bool? ?? true,
        defaultKind: WidgetKind.fromValue(json['defaultKind'] as int?),
        allowPrayerWidget: json['allowPrayerWidget'] as bool? ?? true,
        allowDhikrWidget: json['allowDhikrWidget'] as bool? ?? true,
        theme: WidgetTheme.fromValue(json['theme'] as int?),
        refreshMinutes: json['refreshMinutes'] as int? ?? 30,
        showHijriDate: json['showHijriDate'] as bool? ?? true,
        showCountdown: json['showCountdown'] as bool? ?? true,
        allowBackgroundColor: json['allowBackgroundColor'] as bool? ?? true,
        allowTransparency: json['allowTransparency'] as bool? ?? true,
        allowBackgroundImage: json['allowBackgroundImage'] as bool? ?? true,
        allowCustomWidget: json['allowCustomWidget'] as bool? ?? true,
        customWidgetMaxLength: json['customWidgetMaxLength'] as int? ?? 280,
        defaultWidgetKey: json['defaultWidgetKey'] as String?,
        dhikrCategoryId: json['dhikrCategoryId'] as int?,
        version: json['version'] as int? ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'isEnabled': isEnabled,
        'defaultKind': defaultKind.value,
        'allowPrayerWidget': allowPrayerWidget,
        'allowDhikrWidget': allowDhikrWidget,
        'theme': theme.value,
        'refreshMinutes': refreshMinutes,
        'showHijriDate': showHijriDate,
        'showCountdown': showCountdown,
        'allowBackgroundColor': allowBackgroundColor,
        'allowTransparency': allowTransparency,
        'allowBackgroundImage': allowBackgroundImage,
        'allowCustomWidget': allowCustomWidget,
        'customWidgetMaxLength': customWidgetMaxLength,
        'defaultWidgetKey': defaultWidgetKey,
        'dhikrCategoryId': dhikrCategoryId,
        'version': version,
      };
}
